docs/Secure-Coding-Guide-for-Python/CWE-707/CWE-89/README.md

```diff
@@ -0,0 +1,300 @@
+            return self.cursor.fetchall()[0][0]
+        except sqlite3.OperationalError as operational_error:
+            logging.error(operational_error)
+        return None
+
+    def add_record(self, name: str = ""):
+        """
+        Adds a student name to the table
+            Parameters:
+                name (string): A string with the student name
+        """
+
+        add_values = "INSERT INTO Students(student) VALUES('{name}');"
+        logging.debug("Adding student %s", name)
+        add_values_query = add_values.format(name=name)
+        try:
+            self.cursor.executescript(add_values_query)
+            self.connection.commit()
+        except sqlite3.OperationalError as operational_error:
+            logging.error(operational_error)
+
+
+#####################
+# exploiting above code example
+#####################
+print("sqlite3.sqlite_version=" + sqlite3.sqlite_version)
+
+records = Records()
+records.add_record("Alice")
+records.add_record("Robert'); DROP TABLE students;--")
+records.add_record("Malorny")
+
+print(records.get_record("Alice"))
+# normal as "Robert" has not been added as "Robert":
+print(records.get_record("Robert'); DROP TABLE students;--"))
+print(records.get_record("Malorny"))
+
+```
+
+Adding a student name called `'); DROP TABLE students;--"` allows to drop/delete the table or run any other SQL command.
+
+**Example `noncompliant01.py` output:**
+
+```bash
+sqlite3.sqlite_version=3.34.1
+DEBUG:root:Adding student Alice
+DEBUG:root:Adding student Robert'); DROP TABLE students;--
+DEBUG:root:Adding student Malorny
+ERROR:root:no such table: Students
+ERROR:root:no such table: Students
+None
+ERROR:root:near ")": syntax error
+None
+ERROR:root:no such table: Students
+None
+```
+
+Student `"Malorny"` will never be added because the table `Students` is dropped by the time her parents want to add her.
+
+Static code analysis tools such as `bandit 1.7.4` can be used on the command line to discover or gatekeep vulnerable code anywhere in a product life-cycle:
+
+**Example bandit 1.7.4 output:***
+
+```bash
+Test results:
+>> Issue: [B608:hardcoded_sql_expressions] Possible SQL injection vector through string-based query construction.
+   Severity: Medium   Confidence: Low
+   CWE: CWE-89 (https://cwe.mitre.org/data/definitions/89.html)
+   Location: .\noncompliant01.py:24:21
+   More Info: https://bandit.readthedocs.io/en/1.7.4/plugins/b608_hardcoded_sql_expressions.html
+23              '''
+24              get_values = "SELECT * FROM Students WHERE student = '{name}'".format(name=name)
+25              self.cursor.execute(get_values)
+```
+
+## Compliant Solution - sqlite3
+
+The `compliant01.py` code example is using `sqlite3.cursor.execute(get_values, data_tuple)` that:
+
+* Separates the query from the values to reduce injection attacks.<br>
+`data_tuple = (name,)`<br>
+`get_values = "SELECT * FROM Students WHERE student = ?"`<br>
+`self.cursor.execute(get_values, data_tuple)`<br>
```

Comment:
Perhaps a fenced code block would look better here
````suggestion
* Separates the query from the values to reduce injection attacks.

  ```python
  data_tuple = (name,)
  get_values = "SELECT * FROM Students WHERE student = ?"
  self.cursor.execute(get_values, data_tuple)
  ```
````

Reply:
going to change this